02-library/cs506/sim.py

```
@@ -1,13 +1,52 @@
+import math
+import decimal
+
+
 def euclidean_dist(x, y):
-    raise NotImplementedError()
+    return (p_root(sum(pow(abs(a - b), 2)for a, b in zip(x, y)), 2))
 
 def manhattan_dist(x, y):
-    raise NotImplementedError()
+    return (p_root(sum(pow(abs(a - b), 1)for a, b in zip(x, y)), 1))
 
 def jaccard_dist(x, y):
-    raise NotImplementedError()
+    if not x and not y:
+        return 1
+    return 1 - jaccard_sim(x,y)
 
 def cosine_sim(x, y):
-    raise NotImplementedError()
+    if (len(x) != len(y)):
+        return 'vectors not same size'
+    
+    dotProd = dot(x, y)
+    normX = mag(x)
+    normY = mag(y)
 
+    if (normX * normY) != 0:
+        return dotProd / (normX * normY)
+    else:
+        return float('inf')
```

No! inf is not a valid cosine value. Please raise an error instead